src/main/scala/sangria/gateway/schema/materializer/directive/GraphQLDirectiveProvider.scala

```diff
@@ -31,10 +31,11 @@ class GraphQLDirectiveProvider(implicit ec: ExecutionContext) extends DirectiveP
         c ⇒ {
           val (updatedFields, fragments, vars) = prepareOriginFields(o, c.query, c.schema, c.astFields, c.parentType)
           val varDefs = vars.toVector.flatMap(v ⇒ c.query.operation(c.ctx.operationName).get.variables.find(_.name == v))
-          val queryOp = ast.OperationDefinition(ast.OperationType.Query,
-            name = Some("DelegatedQuery"),
-            variables = varDefs,
-            selections = updatedFields)
+          val (ot, n) = c.query.operations.map(_._2.operationType).toSeq.distinct.toList match {
+            case h :: Nil => (h, Some("Delegated" + h))
+            case x => throw new IllegalStateException(s"Query contains operations [${x.mkString(", ")}].")
+          }
+          val queryOp = ast.OperationDefinition(operationType = ot, name = n, variables = varDefs, selections = updatedFields)
```

Comment:
I think you can replace hardcoded `ast.OperationType.Query` with something like this:

```scala
c.query.operationType(c.ctx.operationName) getOrElse 
  throw new IllegalStateException(s"Can't fins operation type")
```

You need to consider the `operationName` since it defines which operation must be executed if multiple operations are defined in the query.

Otherwise it looks good to me :+1: In future we need to think of a way how we can detach these 2. For example, in case I'm doing a mutation, but the delegate one of the nested fields. in this case delegated query type is `Query` even though original query is a `Mutation`